haskell/test/Stablecoin.hs

```diff
@@ -0,0 +1,140 @@
+
+-- | This function accepts origination params with @Just@ values indicating
+-- the mandatory configuration, and @Nothing@ values for permission parameters
+-- that are not relavant for the test. The return flag indicate if the
+-- contract's permission parameter can meet the required permission configuration.
+originationRequestCompatible :: OriginationParams -> Bool
+originationRequestCompatible op =
+  isPermissionsCompatible (opPermissionsDescriptor op)
+  where
+    isPermissionsCompatible :: PermissionsDescriptorMaybe -> Bool
+    isPermissionsCompatible
+      ( arg #self -> st
+      , arg #pdr ->
+          ( arg #operator -> ot
+          , arg #pdr2 ->
+            ( arg #receiver -> owtmr
+            , arg #pdr3 -> (arg #sender -> otms, arg #custom -> cst)))) = let
+      customFlag = not $ isJust cst
+      selfTransferFlag = case st of
+        Nothing -> True
+        Just (SelfTransferPermitted _) -> True
+        _ -> False
+      operatorTransferFlag = case ot of
+        Nothing -> True
+        Just (OperatorTransferPermitted _) -> True
+        _ -> False
+      owHookSenderFlag = case otms of
+        Nothing -> True
+        Just (OptionalOwnerHook _) -> True
+        _ -> False
+      owHookReceiverFlag = case owtmr of
+        Nothing -> True
+        Just (OptionalOwnerHook _) -> True
+        _ -> False
+      in customFlag && selfTransferFlag && operatorTransferFlag &&
+         owHookSenderFlag && owHookReceiverFlag
+
+mkInitialStorage :: Maybe (TAddress Safelist.Parameter) -> OriginationParams -> Storage
+mkInitialStorage mbSafelistAddr op = let
+  ledgerMap = snd <$> (opBalances op)
+  ledger = #ledger .! (BigMap ledgerMap)
+  mintingAllowances = #minting_allowances .! (BigMap mempty)
+  operators = #operators .! (BigMap operatorMap)
+  owner = #owner .! adminAddr
+  pauser = #pauser .! pauserAddr
+  isPaused = #paused .! False
+  masterMinter = #master_minter .! masterMinterAddr
+  operatorMap = Map.foldrWithKey foldFn mempty (opBalances op)
+  tokenmetadata = #token_metadata .! stablecoinTokenMetadata
+  totalSupply = #total_supply .! (sum $ Map.elems ledgerMap)
+  roles = #roles .! ((masterMinter, owner), (pauser, (#pending_owner_address .! Nothing)))
+  safelistContract = #safelist_contract .! (unTAddress <$> mbSafelistAddr)
+  in ((((ledger, mintingAllowances), (operators, isPaused))
+           , ((stablecoinPermissionsDescriptor, roles), (safelistContract, tokenmetadata))), totalSupply)
+  where
+    foldFn
+      :: Address
+      -> ([Address], Natural)
+      -> Map (Address, Address) ()
+      -> Map (Address, Address) ()
+    foldFn ow (ops, _) m = foldr (\a b -> Map.insert (ow, a) () b) m ops
+
+mkInitialStorageMaybe :: Maybe (TAddress Safelist.Parameter) -> OriginationParams -> Maybe Storage
+mkInitialStorageMaybe mbSafelistAddr op =
+  if originationRequestCompatible op then Just $ mkInitialStorage mbSafelistAddr op
+  else Nothing
+
+specCallbackFA2 :: U.Contract -> Spec
```

Comment:
You should use `origination` function here as was done with `spec_management`

Reply:
Have refactored it.